airflow/contrib/hooks/gcp_dataproc_hook.py

```diff
@@ -33,12 +33,82 @@ def __init__(self, dataproc_api, project_id, job, region='global',
         self.dataproc_api = dataproc_api
         self.project_id = project_id
         self.region = region
+
+        # Check if the job to submit is already running on the cluster.
+        # If so, don't resubmit the job.
+        try:
+            cluster_name = job['job']['placement']['clusterName']
+        except KeyError:
+            self.log.error('Job to submit is incorrectly configured.')
+            raise
+
+        jobs_on_cluster_response = dataproc_api.projects().regions().jobs().list(
+            projectId=self.project_id,
+            region=self.region,
+            clusterName=cluster_name).execute()
+
+        UUID_LENGTH = 9
```

Comment:
This seems to be a bit error prone. How about the following? Let's modify the submit job operator interface by taking a new parameter job_dedupe_regex, when set, the hook will match existing jobs with this job_dedupe_regex and skip a new submission if a new match is found and other conditions are met. 

This way, the hook implementation is entirely decoupled from the operator implementation re: job-name and open to other ways of deduping dataproc jobs.

Reply:
Thank you @fenglu-g. Just to clarify, do you mean that I should modify the [`DataProcHook.submit`](https://github.com/apache/incubator-airflow/blob/0e8394fd23d067b7e226c011bb1825ff734219c5/airflow/contrib/hooks/gcp_dataproc_hook.py#L229) and [`_DataProcJob.__init__`](https://github.com/apache/incubator-airflow/blob/0e8394fd23d067b7e226c011bb1825ff734219c5/airflow/contrib/hooks/gcp_dataproc_hook.py#L31) functions to take in the new regex param like so?

```
class _DataProcJob(LoggingMixin):
    def __init__(self, dataproc_api, project_id, job, region='global',
                 job_error_states=None, job_dedupe_regex=None):

class DataProcHook(GoogleCloudBaseHook):
    def submit(self, project_id, job, region='global', job_error_states=None, job_dedupe_regex=[by default, regex that dedupes by matching the task ID]):
```
And then have `submit()` pass the regex along to the _DataProcJob init [here](https://github.com/apache/incubator-airflow/blob/0e8394fd23d067b7e226c011bb1825ff734219c5/airflow/contrib/hooks/gcp_dataproc_hook.py#L230
), and the deduping happens during the init:

```
def submit(self, project_id, job, region='global', job_error_states=None, job_dedupe_regex=[by default, regex that dedupes by matching the task ID]):
    submitted = _DataProcJob(self.get_conn(), project_id, job, region,                                
 job_error_states=job_error_states, job_dedupe_regex=job_dedupe_regex)
```

Reply:
@fenglu-g Can you take a look at this and let me know if you have any further comments?

Reply:
@fenglu-g Just wanted to clarify if my interpretation of your comment was in line with what you meant? If so I'll go ahead and implement it. Thanks!

Reply:
Sorry for the late reply, yes but we should also surface this job_dedupe_regex argument to dataproc submit job operators (e.g., [DataProcSparkOperator](https://github.com/apache/incubator-airflow/blob/master/airflow/contrib/operators/dataproc_operator.py#L982)), wdyt?

Reply:
Oh gotcha @fenglu-g, so users can provide their own regex (e.g. `DataProcSparkOperator(other args, [my_dedupe_regex_here]`) in `DataprocSparkOperator`, `DataprocPigOperator`, etc. I'll make that change. And sorry about my (very) late reply!

Reply:
Hi @fenglu-g , I started working on the regex idea and I don't think it makes sense to have the user define the deduping logic. 

It seems more error prone to me to have the user define the job ID they want to use for deduping, since the job ID is not user defined, but is instead hardcoded by Airflow here as the task ID + first 8 characters of UUID: https://github.com/apache/incubator-airflow/blob/0e8394fd23d067b7e226c011bb1825ff734219c5/airflow/contrib/hooks/gcp_dataproc_hook.py#L94. Since Airflow sets the job ID, Airflow should also set the deduping logic.

I understand your concern about not wanting to hardcode the deduping logic, though. A solution for this could be: instead of having Airflow hardcode the job ID as in the link above, to have it instead call a function that creates the job ID from the task ID. And then my deduping logic would reference an inverse of that function to extract the task ID for deduping instead of hardcoding the deduping logic. 

In other words, Airflow would create the job ID with a function that maps the task ID -> job ID. And then the dedupe mechanism would call a function that maps from job ID back to task ID to use in deduping. Tests will be added to make sure that these two functions are inverses of each other. 

How does that sound?